File: APP/ios/AirHealthApp/App/EntitlementState.swift
import Foundation

private let entitlementFreshnessWindow: TimeInterval = 24 * 60 * 60

enum VerifiedEntitlementState: String {
    case trialActive = "trial_active"
    case paidActive = "paid_active"
    case expiredReadOnly = "expired_read_only"
}

enum EntitlementFreshness: String {
    case verified = "verified"
    case freshCache = "fresh_cache"
    case staleCache = "stale_cache"
    case inactiveCache = "inactive_cache"
    case missingCache = "missing_cache"
}

enum EffectiveEntitlementMode: String {
    case active = "active"
    case temporaryAccess = "temporary_access"
    case readOnly = "read_only"
}

struct CachedEntitlementSnapshot {
    let sourceState: VerifiedEntitlementState
    let verifiedAtEpochMillis: Int64
}

struct EntitlementCacheState {
    let snapshot: CachedEntitlementSnapshot?
    let isBackendReachable: Bool
    let lastVerificationAttemptAtEpochMillis: Int64?

    static func empty() -> EntitlementCacheState {
        EntitlementCacheState(
            snapshot: nil,
            isBackendReachable: true,
            lastVerificationAttemptAtEpochMillis: nil
        )
    }
}

enum EntitlementCacheAction {
    case snapshotVerified(CachedEntitlementSnapshot)
    case verificationUnavailable(observedAtEpochMillis: Int64)
    case cacheCleared(observedAtEpochMillis: Int64)
}

struct EffectiveEntitlement {
    let mode: EffectiveEntitlementMode
    let freshness: EntitlementFreshness
    let canStartNewSessions: Bool
    let canViewHistory: Bool
    let canEditGoals: Bool
    let canRequestLiveSuggestions: Bool
    let canUseCachedSuggestions: Bool
}

enum EntitlementCacheReducer {
    static func reduce(
        state: EntitlementCacheState,
        action: EntitlementCacheAction
    ) -> EntitlementCacheState {
        switch action {
        case let .snapshotVerified(snapshot):
            return EntitlementCacheState(
                snapshot: snapshot,
                isBackendReachable: true,
                lastVerificationAttemptAtEpochMillis: snapshot.verifiedAtEpochMillis
            )
        case let .verificationUnavailable(observedAtEpochMillis):
            return EntitlementCacheState(
                snapshot: state.snapshot,
                isBackendReachable: false,
                lastVerificationAttemptAtEpochMillis: observedAtEpochMillis
            )
        case let .cacheCleared(observedAtEpochMillis):
            return EntitlementCacheState(
                snapshot: nil,
                isBackendReachable: state.isBackendReachable,
                lastVerificationAttemptAtEpochMillis: observedAtEpochMillis
            )
        }
    }
}

enum EntitlementEvaluator {
    static func deriveEffectiveState(
        state: EntitlementCacheState,
        nowEpochMillis: Int64
    ) -> EffectiveEntitlement {
        guard let snapshot = state.snapshot else {
            return readOnlyEntitlement(freshness: .missingCache)
        }

        if state.isBackendReachable {
            switch snapshot.sourceState {
            case .trialActive, .paidActive:
                return activeEntitlement(freshness: .verified)
            case .expiredReadOnly:
                return readOnlyEntitlement(freshness: .verified)
            }
        }

        let cacheAgeMillis = max(0, nowEpochMillis - snapshot.verifiedAtEpochMillis)
        let hasFreshActiveCache =
            snapshot.sourceState != .expiredReadOnly &&
            TimeInterval(cacheAgeMillis) <= entitlementFreshnessWindow * 1000

        if hasFreshActiveCache {
            return EffectiveEntitlement(
                mode: .temporaryAccess,
                freshness: .freshCache,
                canStartNewSessions: false,
                canViewHistory: true,
                canEditGoals: false,
                canRequestLiveSuggestions: false,
                canUseCachedSuggestions: true
            )
        }

        if snapshot.sourceState == .expiredReadOnly {
            return readOnlyEntitlement(freshness: .inactiveCache)
        }

        return readOnlyEntitlement(freshness: .staleCache)
    }

    private static func activeEntitlement(
        freshness: EntitlementFreshness
    ) -> EffectiveEntitlement {
        EffectiveEntitlement(
            mode: .active,
            freshness: freshness,
            canStartNewSessions: true,
            canViewHistory: true,
            canEditGoals: true,
            canRequestLiveSuggestions: true,
            canUseCachedSuggestions: true
        )
    }

    private static func readOnlyEntitlement(
        freshness: EntitlementFreshness
    ) -> EffectiveEntitlement {
        EffectiveEntitlement(
            mode: .readOnly,
            freshness: freshness,
            canStartNewSessions: false,
            canViewHistory: true,
            canEditGoals: false,
            canRequestLiveSuggestions: false,
            canUseCachedSuggestions: true
        )
    }
}

File: APP/android/app/src/main/java/com/airhealth/app/EntitlementState.kt
package com.airhealth.app

private const val ENTITLEMENT_FRESHNESS_WINDOW_MS = 24L * 60L * 60L * 1000L

enum class VerifiedEntitlementState {
    TRIAL_ACTIVE,
    PAID_ACTIVE,
    EXPIRED,
}

enum class EntitlementFreshness {
    VERIFIED,
    FRESH_CACHE,
    STALE_CACHE,
    INACTIVE_CACHE,
    MISSING_CACHE,
}

enum class EffectiveEntitlementMode {
    ACTIVE,
    TEMPORARY_ACCESS,
    READ_ONLY,
}

data class CachedEntitlementSnapshot(
    val sourceState: VerifiedEntitlementState,
    val verifiedAtEpochMillis: Long,
)

data class EntitlementCacheState(
    val snapshot: CachedEntitlementSnapshot? = null,
    val isBackendReachable: Boolean = true,
    val lastVerificationAttemptAtEpochMillis: Long? = null,
)

sealed interface EntitlementCacheAction {
    data class SnapshotVerified(
        val snapshot: CachedEntitlementSnapshot,
    ) : EntitlementCacheAction

    data class VerificationUnavailable(
        val observedAtEpochMillis: Long,
    ) : EntitlementCacheAction

    data class CacheCleared(
        val observedAtEpochMillis: Long,
    ) : EntitlementCacheAction
}

data class EffectiveEntitlement(
    val mode: EffectiveEntitlementMode,
    val freshness: EntitlementFreshness,
    val canStartNewSessions: Boolean,
    val canViewHistory: Boolean,
    val canEditGoals: Boolean,
    val canRequestLiveSuggestions: Boolean,
    val canUseCachedSuggestions: Boolean,
)

object EntitlementCacheReducer {
    fun reduce(
        state: EntitlementCacheState,
        action: EntitlementCacheAction,
    ): EntitlementCacheState {
        return when (action) {
            is EntitlementCacheAction.SnapshotVerified -> state.copy(
                snapshot = action.snapshot,
                isBackendReachable = true,
                lastVerificationAttemptAtEpochMillis = action.snapshot.verifiedAtEpochMillis,
            )

            is EntitlementCacheAction.VerificationUnavailable -> state.copy(
                isBackendReachable = false,
                lastVerificationAttemptAtEpochMillis = action.observedAtEpochMillis,
            )

            is EntitlementCacheAction.CacheCleared -> EntitlementCacheState(
                snapshot = null,
                isBackendReachable = state.isBackendReachable,
                lastVerificationAttemptAtEpochMillis = action.observedAtEpochMillis,
            )
        }
    }
}

object EntitlementEvaluator {
    fun deriveEffectiveState(
        state: EntitlementCacheState,
        nowEpochMillis: Long,
    ): EffectiveEntitlement {
        val snapshot = state.snapshot
            ?: return readOnlyEntitlement(EntitlementFreshness.MISSING_CACHE)

        val cacheAgeMillis = (nowEpochMillis - snapshot.verifiedAtEpochMillis).coerceAtLeast(0L)
        val isWithinFreshnessWindow = cacheAgeMillis <= ENTITLEMENT_FRESHNESS_WINDOW_MS
        val isActiveSourceState = snapshot.sourceState == VerifiedEntitlementState.TRIAL_ACTIVE ||
            snapshot.sourceState == VerifiedEntitlementState.PAID_ACTIVE

        if (state.isBackendReachable) {
            return when (snapshot.sourceState) {
                VerifiedEntitlementState.TRIAL_ACTIVE,
                VerifiedEntitlementState.PAID_ACTIVE,
                -> if (isWithinFreshnessWindow) {
                    activeEntitlement(EntitlementFreshness.VERIFIED)
                } else {
                    readOnlyEntitlement(EntitlementFreshness.STALE_CACHE)
                }

                VerifiedEntitlementState.EXPIRED -> if (isWithinFreshnessWindow) {
                    readOnlyEntitlement(EntitlementFreshness.VERIFIED)
                } else {
                    readOnlyEntitlement(EntitlementFreshness.INACTIVE_CACHE)
                }
            }
        }

        val hasFreshActiveCache = isActiveSourceState && isWithinFreshnessWindow

        return when {
            hasFreshActiveCache -> EffectiveEntitlement(
                mode = EffectiveEntitlementMode.TEMPORARY_ACCESS,
                freshness = EntitlementFreshness.FRESH_CACHE,
                canStartNewSessions = false,
                canViewHistory = true,
                canEditGoals = false,
                canRequestLiveSuggestions = false,
                canUseCachedSuggestions = true,
            )

            snapshot.sourceState == VerifiedEntitlementState.EXPIRED ->
                readOnlyEntitlement(EntitlementFreshness.INACTIVE_CACHE)

            else -> readOnlyEntitlement(EntitlementFreshness.STALE_CACHE)
        }
    }

    private fun activeEntitlement(freshness: EntitlementFreshness): EffectiveEntitlement {
        return EffectiveEntitlement(
            mode = EffectiveEntitlementMode.ACTIVE,
            freshness = freshness,
            canStartNewSessions = true,
            canViewHistory = true,
            canEditGoals = true,
            canRequestLiveSuggestions = true,
            canUseCachedSuggestions = true,
        )
    }

    private fun readOnlyEntitlement(freshness: EntitlementFreshness): EffectiveEntitlement {
        return EffectiveEntitlement(
            mode = EffectiveEntitlementMode.READ_ONLY,
            freshness = freshness,
            canStartNewSessions = false,
            canViewHistory = true,
            canEditGoals = false,
            canRequestLiveSuggestions = false,
            canUseCachedSuggestions = true,
        )
    }
}
